ddtrace/settings/config.py

```diff
@@ -147,74 +159,56 @@ def is_error_code(self, status_code):
-            del self.tags["service"]
+    service_mapping = En.v(dict, "dd.service.mapping", parser=parse_tags_str, default={})
 
-        # The version tag should not be included on all spans.
-        if self.version and "version" in self.tags:
-            del self.tags["version"]
+    logs_injection = En.v(bool, "dd.logs_injection", default=False)
 
-        self.logs_injection = asbool(os.getenv("DD_LOGS_INJECTION", default=False))
+    report_hostname = En.v(bool, "dd.trace.report_hostname", default=False)
 
-        self.report_hostname = asbool(os.getenv("DD_TRACE_REPORT_HOSTNAME", default=False))
+    health_metrics_enabled = En.v(bool, "dd.trace.health_metrics.enabled", default=False)
 
-        self.health_metrics_enabled = asbool(os.getenv("DD_TRACE_HEALTH_METRICS_ENABLED", default=False))
+    _propagation_style_extract = En.v(
+        set, "dd.trace.propagation_style.extract", parser=_parse_propagation_styles, default={PROPAGATION_STYLE_DATADOG}
+    )
+    _propagation_style_inject = En.v(
+        set, "dd.trace.propagation_style.inject", parser=_parse_propagation_styles, default={PROPAGATION_STYLE_DATADOG}
+    )
 
-        # Propagation styles
-        self._propagation_style_extract = _parse_propagation_styles(
-            "DD_TRACE_PROPAGATION_STYLE_EXTRACT", default=PROPAGATION_STYLE_DATADOG
-        )
-        self._propagation_style_inject = _parse_propagation_styles(
-            "DD_TRACE_PROPAGATION_STYLE_INJECT", default=PROPAGATION_STYLE_DATADOG
-        )
-
-        # Datadog tracer tags propagation
-        x_datadog_tags_max_length = int(os.getenv("DD_TRACE_X_DATADOG_TAGS_MAX_LENGTH", default=512))
-        if x_datadog_tags_max_length < 0 or x_datadog_tags_max_length > 512:
-            raise ValueError(
-                (
-                    "Invalid value {!r} provided for DD_TRACE_X_DATADOG_TAGS_MAX_LENGTH, "
-                    "only non-negative values less than or equal to 512 allowed"
-                ).format(x_datadog_tags_max_length)
-            )
-        self._x_datadog_tags_max_length = x_datadog_tags_max_length
-        self._x_datadog_tags_enabled = x_datadog_tags_max_length > 0
+    _x_datadog_tags_max_length = En.v(
+        int, "DD_TRACE_X_DATADOG_TAGS_MAX_LENGTH", validator=_validate_x_datadog_tags_max_len, default=512
+    )
+    _x_datadog_tags_enabled = En.d(bool, lambda c: c._x_datadog_tags_max_length > 0)
 
-        # Raise certain errors only if in testing raise mode to prevent crashing in production with non-critical errors
-        self._raise = asbool(os.getenv("DD_TESTING_RAISE", False))
-        self._trace_compute_stats = asbool(os.getenv("DD_TRACE_COMPUTE_STATS", False))
-        self._appsec_enabled = asbool(os.getenv("DD_APPSEC_ENABLED", False))
+    _raise = En.v(bool, "dd.testing.raise", default=False)
+    _trace_compute_stats = En.v(bool, "dd.trace.compute_stats", default=False)
 
     def __getattr__(self, name):
-        if name not in self._config:
```

Comment:
🤔  I believe that `__getattr__` is called _conditionally_ if the attribute is not found. Since we are setting the attribute when we "miss", the next time around the previously created instance of `IntegrationConfig` is returned.

Reply:
I imagine it could happen in theory. So probably be safer here.
Though I imagine the problem already exists in the current code?